### ISSUE TITLE
[CLOSED] Port to electron + add new features

### ISSUE THREAD

**Issue by desplesda** [YarnSpinnerTool/YarnEditor#68]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [blurymind](https://github.com/blurymind)**
_Sunday Jun 10, 2018 at 19:20 GMT_
_Originally opened as https://github.com/InfiniteAmmoInc/Yarn/pull/68_

----

Hi, this is somewhat of a big change pull, so please bear with me

It ports Yarn over to electron, while also adding a couple of new features:

- Ability to easily create links to other existing nodes via a new helper menu:
 
![yarn-linkmaking](https://user-images.githubusercontent.com/6495061/41685764-7bf48d1a-74d9-11e8-89bc-b7bae39470f6.gif)
(addressing #34 )

- Ability to test dialogue via e new "Run" menu (using bondage.js and bbcode)
 
![yarn-testerbbcode](https://user-images.githubusercontent.com/6495061/41685950-2b8b3580-74da-11e8-89ea-c7d23dea19da.gif)
(addressing #29 )

- file drag and drop (including support for dropping multiple files )
![yarn-dropfile](https://user-images.githubusercontent.com/6495061/41766680-8ce4ad2c-75ff-11e8-950e-f35346769c97.gif)
(addressing #32 )

- Automatically create New Nodes from [[new]] and [[new|choice]] - if no nodes exist for them:
![yarn-autonewnodefromchoice](https://user-images.githubusercontent.com/6495061/41983609-8f78d286-7a26-11e8-916b-d6272d184843.gif)
(addressing https://github.com/InfiniteAmmoInc/Yarn/issues/27)

- Ability to run more than one instance of Yarn
(addressing #59 )

- Lots of fixes and new key bindings

I am not sure if the shift to electron would be welcome by the original creators:
@InfiniteAmmoInc , @seiyria and @beeglebug

Please let me know,so I know if a better route for me is to fork it to "Yarn-plus" or "Yarn-electron" or something

In order to try it just run
npm install
then
npm start

----

_**[blurymind](https://github.com/blurymind)** included the following code: https://github.com/InfiniteAmmoInc/Yarn/pull/68/commits_

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Monday Jun 11, 2018 at 22:55 GMT_

----

@charblar hi thank you for looking into this. Sorry I forgot to mention that bondage.js doesnt support running the syntax 
`<<function(parameters)>>`
At the moment. I have submited another pull request to add that already here:
https://github.com/jhayley/bondage.js/pull/37
And used the same gif to demonstrate it there

The author of bondage.js @jhayley will review it when she has time, but basically adds the feature.
Feel free to run the test with it and see if it works :)

I am going to work on this a bit more this week hopefully to clear up some corner cases where it spills an error upon completion and center the text window. The code can do with some cleaning up too :)
I have to fix the styling on one of the search filters

**Comment by desplesda** [comment]
<a href="https://github.com/charblar"><img src="https://avatars0.githubusercontent.com/u/25385583?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [charblar](https://github.com/charblar)**
_Tuesday Jun 12, 2018 at 11:42 GMT_

----

@blurymind hey, no problem! I assumed that was the case. I personally know Hayley and will work with her in the coming days to resolve that.

Thank you for the clarification, looking forward to your additional commits!

Edit: Oops hit comment and close, damn you GitHub and your lack of confirmation messages.

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Tuesday Jun 12, 2018 at 19:37 GMT_

----

@charblar thank you for looking into that. It wasn't really intended to be a part of this pull request - and still isnt. I needed the javascript parsing capability for my personal project. So in my pull request here that is not supported, it's more of a feature for bondage.js and not yarn.

On the front of this one - I fixed the previous styling problems and console errors, while also cleaning up the code.Now it should work much better :) 

I am now shifting my focus on this pull and will go back to bondage.js a bit later.  @jhayley doesnt like the syntax `<<jsfunction(parameters)>>` , so I might have to change it to `<<eval jsfunction(parameters)>>` please join the discussion at the other git repo pull if you have suggestions/ideas :)

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jun 13, 2018 at 09:27 GMT_

----

Ok going back to bondage.js, my approach was not right, so I will create another pull request - possibly later in the week for adding support to do <<command(parameters)>> in bondage.js.
I misunderstood the intended design of <<>> and the problem, so scrapped it and will redo that to allow <<function(parameters)>> syntax without introducing a new node type

For the commands - maybe later on I can add to the story tester the ability to show commands that were passed to it as well. I have noticed that it was kind of requested before

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Thursday Jun 14, 2018 at 19:48 GMT_

----

A little update on this- after reading some requests here I decided to add the option to also show commands like this:
![yarnrunnerupdate1](https://user-images.githubusercontent.com/6495061/41434606-3936d20e-7014-11e8-8903-17dc6575691b.gif)
Also made the window self close when finished

I have to re-enable some things and sort them out before porting is complete. I also want to add one more usability feature to Yarn

**Comment by desplesda** [comment]
<a href="https://github.com/charblar"><img src="https://avatars0.githubusercontent.com/u/25385583?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [charblar](https://github.com/charblar)**
_Friday Jun 15, 2018 at 22:45 GMT_

----

Awesome, thanks for keeping us updated. Looking forward to it.

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Friday Jun 15, 2018 at 22:56 GMT_

----

Another update on this- I re-enabled all the file commands now- so the electron version is on-par with the main one.
Please note that the command debugging depends on a pull request I made to bondage.js - so command debugging is not there yet, since that needs to be merged first

I also made some changes to my vscode start script to get it to run it from a portable flash thumb- if you are trying it- you might have to revert the changes in .vscode/tasks.json for now :) 

Anyways, I hope to have an electron build ready for testing soon-ish

**Comment by desplesda** [comment]
<a href="https://github.com/hylyh"><img src="https://avatars2.githubusercontent.com/u/1892337?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [hylyh](https://github.com/hylyh)**
_Friday Jun 15, 2018 at 23:57 GMT_

----

As a note, the changes being made to bondage.js are largely unrelated to the electron change and i recommend separating these concerns. Those feature requests aren't needed to port yarn to electron

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Saturday Jun 16, 2018 at 00:13 GMT_

----

@jhayley yes of course. I have been using the electron port for testing the pull I am doing to bondage.js

Just wanted to note that the electron port at the moment will have this feature broken for most

Edit:
I have removed it from the electron port

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jun 20, 2018 at 21:39 GMT_

----

A little update on this.
There is now an electron release- that removes the command debugging for now and uses the official bondage.js branch (@jhayley )
:) 
You can download it here
https://github.com/blurymind/Yarn/releases/tag/untagged-1933231011a749959b56

@charblar removing command debugging resolves the change you requested :) 

For next update I am going to implement mid-click panning and try to also make linux and mac builds.
Another feature I am adding is automatic updating via the electron framework

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Friday Jun 22, 2018 at 19:33 GMT_

----

A little update:

I added:
- file drag and drop (including support for dropping multiple files )
![yarn-dropfile](https://user-images.githubusercontent.com/6495061/41766680-8ce4ad2c-75ff-11e8-950e-f35346769c97.gif)

dropping multiple files appends them:
![yarn-dropfilesmany](https://user-images.githubusercontent.com/6495061/41774372-6978e9da-7617-11e8-8860-3f0189abd4aa.gif)

- Support for deleting multiple selected nodes (via the delete key)
- Fix spacebar key (should be on key release- so it doesnt repeat while held)
- new key binding- hitting enter (after switching to a node via spacebar) will open it
- new key binding- ctrl+space while editing a node will switch to edit the next node
- new key binding- ctrl+enter while editing a node will save and close the node
- run document format prettify

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jun 27, 2018 at 15:30 GMT_

----

A little update today - I added another usability feature 👍  
![yarn-autonewnodefromchoice](https://user-images.githubusercontent.com/6495061/41983609-8f78d286-7a26-11e8-916b-d6272d184843.gif)
:) 
This addresses #27 
I will merge the commit tonight

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jun 27, 2018 at 15:39 GMT_

----

@charblar this is shaping up to be an epic release and I have addressed your request - it is no longer spitting out an error :) 
For now I am not going to add the command debugger to the tester

Can you tell me what else needs to be done to get this merged?

**Comment by desplesda** [comment]
<a href="https://github.com/julsam"><img src="https://avatars3.githubusercontent.com/u/340056?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [julsam](https://github.com/julsam)**
_Wednesday Jun 27, 2018 at 17:19 GMT_

----

I really love the new features @blurymind +1 , but (and I don't want to sound like an ungrateful ***,) i wished you didn't run prettify... or at least not in the same commit as other changes.

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jun 27, 2018 at 17:24 GMT_

----

@julsam  sorry- its a habbit from another project. I can undo that if @charblar requests it :)

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jun 27, 2018 at 22:09 GMT_

----

New build out - go grab a copy! :) 
https://github.com/blurymind/Yarn/releases/tag/untagged-4a6d4d2b95bf71bdd264

**Comment by desplesda** [comment]
<a href="https://github.com/charblar"><img src="https://avatars0.githubusercontent.com/u/25385583?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [charblar](https://github.com/charblar)**
_Sunday Jul 01, 2018 at 21:35 GMT_

----

@blurymind Hey, I was out of country the past week and am taking a look at all of this now.  Thank you for you patience. 

For the remainder of this PR I request that we hold off on features and changes outside the scope of porting to Electron.

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Sunday Jul 01, 2018 at 23:42 GMT_

----

@charblar Hi no problem. I will stop adding any new features and focus on the porting.

From my tests the electron port seems to work fine. Can you please test and review it :) I hope I haven't introduced any regressions

**Comment by desplesda** [comment]
<a href="https://github.com/charblar"><img src="https://avatars0.githubusercontent.com/u/25385583?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [charblar](https://github.com/charblar)**
_Monday Jul 02, 2018 at 12:59 GMT_

----

@blurymind Looks good to me, I would ask that you first ignore the .vscode directory first and remove the existing files from inside of it and then we can merge it in.

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Monday Jul 02, 2018 at 17:06 GMT_

----

> @blurymind Looks good to me, I would ask that you first ignore the .vscode directory first and remove the existing files from inside of it and then we can merge it in.

@charblar It is done now :) 
Thank you for reviewing the changes.

**Comment by desplesda** [comment]
<a href="https://github.com/julsam"><img src="https://avatars3.githubusercontent.com/u/340056?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [julsam](https://github.com/julsam)**
_Wednesday Jul 04, 2018 at 19:05 GMT_

----

Just to add more feedback: I've been using a modified version of @blurymind 's electron port for the last 2 weeks, I have not encounter any problems so far, it's very stable.

I've been using it only on windows though (win7 x64).

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jul 04, 2018 at 19:19 GMT_

----

@julsam thank you for taking the time to test it. I noticed that you even merged my changes to your fork (https://github.com/julsam/VineEditor) which is flattering 😊 
After @charblar merges these changes I think of adding a few more small but useful things in a separate pull.

**Comment by desplesda** [comment]
<a href="https://github.com/blurymind"><img src="https://avatars0.githubusercontent.com/u/6495061?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [blurymind](https://github.com/blurymind)**
_Wednesday Jul 04, 2018 at 20:42 GMT_

----

@charblar  Thank you !! 🎉 😊 🎉 🍾
